worker/codalabworker/worker.py

```diff
@@ -385,3 +390,226 @@ def _upgrade(self):
+                # Failed to download dependency; -> FINALIZING
+                run_state.info['failure_message'] = 'Failed to download dependency %s: %s' % (
+                        dep['child_path'], '') #TODO: get more specific message
+                return run_state._replace(stage=RunStage.FINALIZING, info=run_state.info)
+
+        docker_image = run_state.resources['docker_image']
+        image_state = self._image_manager.get(docker_image)
+        if image_state.stage == DependencyStage.DOWNLOADING:
+            status_message = 'Pulling docker image: ' + (image_state.message or docker_image)
+            return run_state._replace(run_status=status_message)
+        elif image_state.stage == DependencyStage.FAILED:
+            # Failed to pull image; -> FINALIZING
+            run_state.info['failure_message'] = image_state.message
+            return run_state._replace(stage=RunStage.FINALIZING, info=run_state.info)
+
+        # All dependencies ready! Set up directories, symlinks, container. Start container.
+        # 1) Set up a directory to store the bundle.
+        remove_path(run_state.bundle_path)
+        os.mkdir(run_state.bundle_path)
+
+        # 2) Set up symlinks
+        bundle_uuid = run_state.bundle['uuid']
+        dependencies = []
+        docker_dependencies_path = '/' + bundle_uuid + '_dependencies'
+        for dep in run_state.bundle['dependencies']:
+            child_path = os.path.normpath(os.path.join(run_state.bundle_path, dep['child_path']))
+            if not child_path.startswith(run_state.bundle_path):
+                raise Exception('Invalid key for dependency: %s' % (dep['child_path']))
+
+            dependency_path = self._dependency_manager.get(
+                        (dep['parent_uuid'], dep['parent_path'])).path
+            dependency_path = os.path.join(self._bundles_dir, dependency_path)
+
+            docker_dependency_path = os.path.join(docker_dependencies_path, dep['child_path'])
+            os.symlink(docker_dependency_path, child_path)
+            dependencies.append((dependency_path, docker_dependency_path))
+
+        # 3) Set up container
+        if run_state.resources['request_network']:
+            docker_network = self.docker_network_external_name
+        else:
+            docker_network = self.docker_network_internal_name
+
+        cpuset, gpuset = self._assign_cpu_and_gpu_sets(
+                run_state.resources['request_cpus'], run_state.resources['request_gpus'])
+
+        # 4) Start container
+        container_id = self._docker.start_container(
+            run_state.bundle_path, bundle_uuid, run_state.bundle['command'],
+            run_state.resources['docker_image'], docker_network, dependencies,
+            cpuset, gpuset, run_state.resources['request_memory']
+        )
+
+        digest = self._docker.get_image_repo_digest(run_state.resources['docker_image'])
+        #self._image_manager.touch_image(digest)
```

Comment:
Why is this commented out?

Reply:
because this function has not yet been (properly) implemented